src/kernel/heap.zig

```diff
@@ -213,25 +230,51 @@ pub const Heap = struct {
         const search_result = NodeSearch.search(addr, addr_end, 0, self.size) orelse unreachable;
 
         const entry_start = search_result.start / self.block_min_size;
-        const entry_end = search_result.end / self.block_min_size;
+        const entry_end = search_result.end / self.block_min_size + 1;
 
         // Clear entries associated with the order that the allocation was stored in
-        var entry: u32 = entry_start;
-        while (entry < entry_end) : (entry += 1) {
+        var entry: u32 = @intCast(u32, entry_start);
+        while (entry < entry_end and entry < self.size / self.block_min_size) : (entry += 1) {
             self.bitmap.clearEntry(entry) catch unreachable;
         }
     }
 };
 
+///
+/// Initialise a heap to keep track of allocated memory.
+///
+/// Arguments:
+///     IN vmm_payload: type - The virtual memory manager's payload type.
+///     IN heap_vmm: vmm.VirtualMemoryManager(vmm_payload) - The virtual memory manager that will allocate a region of memory for the heap to govern.
+///     IN attributes: vmm.Attributes - The attributes to apply to the heap's memory.
+///     IN heap_size: usize - The size of the heap. Must be greater than zero, a power of two and should be a multiple of the vmm's block size.
+///     IN allocator: *std.mem.Allocator - The allocator to use to initialise the heap structure.
+///
+/// Return: Heap
+///     The heap constructed.
+///
+/// Error: std.mem.Allocator.Error || Error
+///     std.mem.Allocator.Error.OutOfMemory: There wasn't enough memory in the allocator to create the heap.
+///     Heap.Error.NotPowerOfTwo: The heap size isn't a power of two or is 0.
+///
+pub fn init(comptime vmm_payload: type, heap_vmm: *vmm.VirtualMemoryManager(vmm_payload), attributes: vmm.Attributes, heap_size: usize, allocator: *Allocator) (Allocator.Error || Error)!Heap {
+    log.logInfo("Init heap\n", .{});
+    defer log.logInfo("Done heap\n", .{});
+    var heap_start = (try heap_vmm.alloc(heap_size / vmm.BLOCK_SIZE, attributes)) orelse panic(null, "Not enough contiguous physical memory blocks to allocate to kernel heap\n", .{});
+    return try Heap.init(heap_start, heap_size, 16, allocator);
```

Comment:
Could add an errdefer to free the VMM memory allocated if the heap returns an error

Reply:
Good spot